### PR TITLE
Master mcvijetinovic implementing test suite junit and mockito

### PR DESCRIPTION
**Pull request by mlcvijetinovic**
Implementation of jUnit&Mock test cases

- Added new EntityBuilder class for easy creating test objects
- Added test examples for user service functionality

### CODE REVIEW

File: src/main/java/eu/execom/labs/teamodoro/entity/EntityBuilder.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+        user.setId(1l);
+        user.setGoogleId("");
+        user.setEmail("");
+        user.setName("");
+        return user;
+    }
+
+    public static Team team() {
+        Team team = new Team();
+        team.setName("");
```
Comment (stanislavgrujic) [inline src/main/java/eu/execom/labs/teamodoro/entity/EntityBuilder.java]: Please set a name other than an empty string.

File: src/main/java/eu/execom/labs/teamodoro/entity/EntityBuilder.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+public class EntityBuilder {
+
+    public static User user() {
+        User user = new User();
+        user.setId(1l);
+        user.setGoogleId("");
+        user.setEmail("");
+        user.setName("");
+        return user;
```
Comment (stanislavgrujic) [inline src/main/java/eu/execom/labs/teamodoro/entity/EntityBuilder.java]: Please set some values other than an empty string from user properties.

File: src/main/java/eu/execom/labs/teamodoro/entity/EntityBuilder.java [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+        userTeam.setUser(user);
+        userTeam.setTeam(team);
+        userTeam.setId(userTeamId());
+        return userTeam;
+    }
+
+    public static UserTeamId userTeamId() {
+        UserTeamId userTeamId = new UserTeamId();
+        userTeamId.setUserId(user().getId());
+        userTeamId.setTeamId(team().getId());
```
Comment (stanislavgrujic) [inline src/main/java/eu/execom/labs/teamodoro/entity/EntityBuilder.java]: Team doesn't have an id set, so you will have null here.

File: src/test/java/eu/execom/labs/temodoro/service/TestUserService.java [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+import eu.execom.labs.teamodoro.entity.User;
+import eu.execom.labs.teamodoro.entity.UserTeam;
+import eu.execom.labs.teamodoro.repository.TeamRepository;
+import eu.execom.labs.teamodoro.repository.UserRepository;
+import eu.execom.labs.teamodoro.repository.UserTeamRepository;
+import eu.execom.labs.teamodoro.service.TeamService;
+import eu.execom.labs.teamodoro.service.UserService;
+
+@RunWith(MockitoJUnitRunner.class)
+public class TestUserService {
```
Comment (stanislavgrujic) [inline src/test/java/eu/execom/labs/temodoro/service/TestUserService.java]: Common practice is to name a test class ass *_ClassBeingTested_Test*, so in this case would be *UserServiceTest*.

File: src/test/java/eu/execom/labs/temodoro/service/TestUserService.java [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+    private UserRepository userRepository;
+
+    @Mock
+    private TeamRepository teamRepository;
+
+    @Mock
+    private UserTeamRepository userTeamRepository;
+
+    @InjectMocks
+    UserService userService;
```
Comment (stanislavgrujic) [inline src/test/java/eu/execom/labs/temodoro/service/TestUserService.java]: Please make this field private.

File: src/test/java/eu/execom/labs/temodoro/service/TestUserService.java [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+
+    @InjectMocks
+    private TeamService teamService;
+
+    @Test
+    public void shouldFindUserWithGoogleIdSuccessfully() {
+        // given
+        User existingUser;
+        String googleId = "123";
+        when(userRepository.findByGoogleId(googleId)).thenReturn(EntityBuilder.user());
```
Comment (stanislavgrujic) [inline src/test/java/eu/execom/labs/temodoro/service/TestUserService.java]: You should extract *EntityBuilder.user()* to a variable and to do following:
```
User expected = EntityBuilder.user();
when(userRepository.findByGoogleId(googleId)).thenReturn(expected);

/--- other code ---/

assertEquals(expected, existingUser);
```

File: src/test/java/eu/execom/labs/temodoro/service/TestUserService.java [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+
+        // when
+        existingUser = userService.findUserByGoogleId(googleId);
+
+        // then
+        assertNotNull(existingUser);
+        assertEquals(EntityBuilder.user(), existingUser);
+        verify(userRepository).findByGoogleId(googleId);
+
+    }
```
Comment (stanislavgrujic) [inline src/test/java/eu/execom/labs/temodoro/service/TestUserService.java]: No need to add an empty line before method end. Please remove on all methods in the class.

File: src/test/java/eu/execom/labs/temodoro/service/TestUserService.java [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+        verify(userRepository).save(user);
+
+    }
+
+    @Test
+    public void shouldFailUserSaveWhenNullValueIsPassed() {
+        // given
+        User user = EntityBuilder.user();
+        User existingUser;
+        when(userRepository.save(user)).thenReturn(null);
```
Comment (stanislavgrujic) [inline src/test/java/eu/execom/labs/temodoro/service/TestUserService.java]: You are actually not passing null value. Please think and refactor the test.

File: src/test/java/eu/execom/labs/temodoro/service/TestUserService.java [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+        List<User> allUsers = new ArrayList<>();
+        allUsers.add(EntityBuilder.user());
+        when(userRepository.findAll()).thenReturn(allUsers);
+
+        // when
+        List<User> listUser = userService.findAllUsers();
+
+        // then
+        assertFalse("List is not empty", listUser.isEmpty());
+        verify(userRepository, times(1)).findAll();
```
Comment (stanislavgrujic) [inline src/test/java/eu/execom/labs/temodoro/service/TestUserService.java]: *times(1)* is default value, so you can omit it here.

File: src/test/java/eu/execom/labs/temodoro/service/TestUserService.java [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+    private TeamRepository teamRepository;
+
+    @Mock
+    private UserTeamRepository userTeamRepository;
+
+    @InjectMocks
+    UserService userService;
+
+    @InjectMocks
+    private TeamService teamService;
```
Comment (stanislavgrujic) [inline src/test/java/eu/execom/labs/temodoro/service/TestUserService.java]: Unused field. Please remove it. It does not belong here at the first place.

File: src/test/java/eu/execom/labs/temodoro/service/TestUserService.java [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+        // then
+        assertFalse("List is not empty", listUser.isEmpty());
+        verify(userRepository, times(1)).findAll();
+    }
+
+    @Test
+    public void shouldFindAllTeamsSuccessfully() {
+        // given
+        User user = EntityBuilder.user();
+        user.setId(1L);
```
Comment (stanislavgrujic) [inline src/test/java/eu/execom/labs/temodoro/service/TestUserService.java]: This is already set internally in EntityBuilder.

File: src/main/java/eu/execom/labs/teamodoro/entity/EntityBuilder.java (hunk 3 above)
Comment (mlcvijetinovic) [inline src/main/java/eu/execom/labs/teamodoro/entity/EntityBuilder.java]: Fixed this,by setting team id.

**Comment by stanislavgrujic** [comment]
Please update branch with the latest master version.

**Comment by mlcvijetinovic** [comment]
I update branch with new master,and commited new changes.

File: src/main/java/eu/execom/labs/teamodoro/entity/EntityBuilder.java [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+
+    public static UserTeam userTeam() {
+        UserTeam userTeam = new UserTeam();
+        userTeam.setUser(user());
+        userTeam.setTeam(team());
+        userTeam.setId(userTeamId(userTeam.getUser(), userTeam.getTeam()));
+        return userTeam;
+    }
+
+    public static UserTeam userTeamWithParam(User user, Team team) {
```
Comment (stanislavgrujic) [inline src/main/java/eu/execom/labs/teamodoro/entity/EntityBuilder.java]: Please use the benefits of method overloading - rename the method to *userTeam*

File: src/test/java/eu/execom/labs/temodoro/service/TestUserService.java [hunk 13]
```diff
@@ -0,0 +1,205 @@
...
+    }
+
+    @Test
+    public void shouldFindUserWithIdSuccessfully() {
+        // given
+        User existingUser = EntityBuilder.user();
+        when(userRepository.findOne(existingUser.getId())).thenReturn(existingUser);
+
+        // when
+        User expectingUser = userService.findById(existingUser.getId());
```
Comment (stanislavgrujic) [inline src/test/java/eu/execom/labs/temodoro/service/TestUserService.java]: Please rename to *expectedUser*.

File: src/test/java/eu/execom/labs/temodoro/service/TestUserService.java [hunk 14]
```diff
@@ -0,0 +1,205 @@
...
+    }
+
+    @Test
+    public void shouldSaveUserSuccessfully() {
+        // given
+        User existingUser = EntityBuilder.user();
+        when(userRepository.save(existingUser)).thenReturn(existingUser);
+
+        // when
+        User expectingUser = userService.save(existingUser);
```
Comment (stanislavgrujic) [inline src/test/java/eu/execom/labs/temodoro/service/TestUserService.java]: Please rename to expectedUser.

File: src/test/java/eu/execom/labs/temodoro/service/TestUserService.java [hunk 15]
```diff
@@ -0,0 +1,205 @@
...
+    @Test
+    public void shouldReturnAllUsersSuccessfully() {
+        // given
+        User existingUser = EntityBuilder.user();
+        List<User> allUsers = new ArrayList<>();
+        allUsers.add(existingUser);
+        when(userRepository.findAll()).thenReturn(allUsers);
+
+        // when
+        List<User> listUser = userService.findAllUsers();
```
Comment (stanislavgrujic) [inline src/test/java/eu/execom/labs/temodoro/service/TestUserService.java]: It is not a good practice to name a variable with type included - please rename to *users*.

File: src/test/java/eu/execom/labs/temodoro/service/TestUserService.java [hunk 16]
```diff
@@ -0,0 +1,205 @@
...
+        // given
+        User existingUser = EntityBuilder.user();
+        Team existingTeam = EntityBuilder.team();
+        Set<UserTeam> userTeam = new LinkedHashSet<>();
+        userTeam.add(EntityBuilder.userTeamWithParam(existingUser, existingTeam));
+        existingUser.setUserTeams(userTeam);
+        when(userRepository.findOne(existingUser.getId())).thenReturn(existingUser);
+
+        // when
+        List<Team> teamList = userService.findAllTeams(existingUser.getId());
```
Comment (stanislavgrujic) [inline src/test/java/eu/execom/labs/temodoro/service/TestUserService.java]: Please rename to teams. Also, please try to be as consistent as possible - in the previous method you have used the name *listUser* and here you are using *teamList*.

File: src/main/java/eu/execom/labs/teamodoro/entity/EntityBuilder.java (hunk 12 above)
Comment (mlcvijetinovic) [inline src/main/java/eu/execom/labs/teamodoro/entity/EntityBuilder.java]: Fixed all commented code below with last commit,but comment isnt closed.

**Comment by stanislavgrujic** [comment]
Please resolve merge conflicts.

**Pull request merged**